man/ncsdis.1
.Dd October 28, 2015
.Dt NCSDIS 1
.Os
.Sh NAME
.Nm ncsdis
.Nd BioWare NWScript bytecode disassembler
.Sh SYNOPSIS
.Nm ncsdis
.Op Ar options
.Ar input_file
.Op Ar output_file
.Sh DESCRIPTION
.Nm
disassembles NCS files, compiled bytecode of the NWScript scripting
language used by every single Aurora engine game, except for the
Nintendo DS game
.Em Sonic Chronicles: The Dark Brotherhood .
The output, while not as clear as the C-like source code that produced
the bytecode, is human-readable.
.Pp
.Nm
is meant as a better maintainable and more complete replacement
of the disassembly mode found in the OpenKnights nwnnsscomp tool.
As such, it fixes provides these enhancements:
.Bl -bullet -compact
.It
Compilable in 64-bit mode
.It
Out of the box support for all Aurora engine games
.It
Support for array opcodes added in
.Em Dragon Age: Origins
.It
Support for reference opcodes added in
.Em Dragon Age II
.It
Knowledge of additional variable types
.It
No need to provide an external nwscript.nss file
.El
.Pp
Additionally,
.Nm
can create a dot file specifying a control flow
graph of the disassembly.
It can be plotted into a graph by the
.Xr dot 1
tool of the GraphViz suite.
.Pp
.Sy WARNING :
these graphs can get very large very quickly.
To keep the resulting image small,
set a small font and font size when calling GraphViz,
and decrease the nodesep and ranksep values.
.Pp
Since there is no way to automatically detect for which game this
script is, this information must be provided on the command.
If no game is specified, the ACTION opcode that call an engine function,
such as
.Dq GetModule
or trigonometry functions, will only display a
number instead of a function name.
.Sh OPTIONS
.Bl -tag -width xxxx -compact
.It Fl h
.It Fl Fl help
Show a help text and exit.
.It Fl Fl version
Show version information and exit.
.It Fl Fl list
Create a full disassembly listing, including byte addresses and the
raw bytecode, similar to the disassembly mode of nwnnsscomp.
This is the default mode.
.It Fl Fl assembly
Only create the human-readable mnemonics part of the disassembly
listing.
Useful for modifying and extending the script.
.It Fl Fl dot
Create a flow control graph in the dot language, to be plotted by
the GraphViz suite.
.It Fl Fl stack
Print the stack frame for each instruction.
Only available in list or assembly mode, not in dot mode.
.It Fl Fl control
Print detected control structures inside block nodes.
Only available in dot mode.
.It Fl Fl nwn
Use engine function tables of the game
.Em Neverwinter Nights .
.It Fl Fl nwn2
Use engine function tables of the game
.Em Neverwinter Nights 2 .
.It Fl Fl kotor
Use engine function tables of the game
.Em Star Wars: Knights of the Old Republic .
.It Fl Fl kotor2
Use engine function tables of the game
.Em Star Wars: Knights of the Old Republic II \(en The Sith Lords .
.It Fl Fl jade
Use engine function tables of the game
.Em Jade Empire .
.It Fl Fl witcher
Use engine function tables of the game
.Em The Witcher .
.It Fl Fl dragonage
Use engine function tables of the game
.Em Dragon Age: Origins .
.It Fl Fl dragonage2
Use engine function tables of the game
.Em Dragon Age II .
.El
.Bl -tag -width xxxx -compact
.It Ar input_file
The NCS file to disassemble.
.It Ar output_file
The disassembly will be written there.
If no output file is specified, the disassembly will be written to
.Dv stdout .
.El
.Sh EXAMPLES
Disassemble the script
.Pa file.ncs :
.Pp
.Dl $ ncsdis file.ncs file.lst
.Pp
Disassemble the script
.Pa file.ncs
and write the listing to
.Dv stdout :
.Pp
.Dl $ ncsdis file.ncs
.Pp
Disassemble the Jade Empire script
.Pa file.ncs :
.Pp
.Dl $ ncsdis --jade file.ncs file.list
.Pp
Disassemble the Jade Empire script
.Pa file.ncs
in pure disassembly mode:
.Pp
.Dl $ ncsdis --assembly --jade file.ncs file.asm
.Pp
Create a dot graph file of the Neverwinter Nights script
.Pa file.ncs :
.Pp
.Dl $ ncsdis --dot --nwn file.ncs file.dot
.Pp
Create a dot graph file of the Neverwinter Nights script
.Pa file.ncs
and plot it:
.Bd -literal -offset xxxxxx
$ ncsdis --dot --nwn file.ncs | dot -Gnodesep=0.1 -Granksep=0.1 \e
  -Gfontname="Courier New" -Nfontname="Courier New" -Gfontsize=10 \e
  -Nfontsize=8 -Earrowsize=0.5 -Tpng > file.png
.Ed
.Sh SEE ALSO
.Xr dot 1 ,
.Xr nwnnsscomp 1
.Pp
More information about the xoreos project can be found on
.Lk https://xoreos.org/ "its website"
.Ns .
.Sh AUTHORS
This program is part of the xoreos-tools package, which in turn is
part of the xoreos project, and was written by the xoreos team.
Please see the
.Pa AUTHORS
file for details.


NEWS
Changes in xoreos-tools version 0.0.2
=====================================

This is the first official release of xoreos-tools, together with xoreos.

Included are:

- Tools for handling resource archives
  - unkeybif
  - unerf
  - unrim
  - unherf
  - unnds
  - desmall
- Tools converting basic resource formats into human-readable form
  - gff2xml
  - tlk2xl
  - convert2da
- Tools converting graphics formats into TGA
  - xoreostex2tga
  - unnsbtx
  - nbfs2tga
  - ncgr2tga
  - cbgt2tga
  - cdpth2tga


src/ncsdis.cpp
/* xoreos-tools - Tools to help with xoreos development
 *
 * xoreos-tools is the legal property of its developers, whose names
 * can be found in the AUTHORS file distributed with this source
 * distribution.
 *
 * xoreos-tools is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 3
 * of the License, or (at your option) any later version.
 *
 * xoreos-tools is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with xoreos-tools. If not, see <http://www.gnu.org/licenses/>.
 */

/** @file
 *  Tool to disassemble NWScript bytecode.
 */

#include <cassert>
#include <cstring>
#include <cstdio>

#include <vector>

#include "src/version/version.h"

#include "src/common/scopedptr.h"
#include "src/common/ustring.h"
#include "src/common/util.h"
#include "src/common/error.h"
#include "src/common/platform.h"
#include "src/common/readfile.h"
#include "src/common/writefile.h"
#include "src/common/stdoutstream.h"
#include "src/common/cli.h"

#include "src/aurora/types.h"

#include "src/nwscript/disassembler.h"

#include "src/util.h"

enum Command {
	kCommandNone     = -1,
	kCommandListing  =  0,
	kCommandAssembly =  1,
	kCommandDot      =  2,
	kCommandNSS      =  3,
	kCommandMAX
};

bool parseCommandLine(const std::vector<Common::UString> &argv, int &returnValue,
                      Common::UString &inFile, Common::UString &outFile,
                      Aurora::GameID &game, Command &command,
                      bool &printStack, bool &printControlTypes);

void disNCS(const Common::UString &inFile, const Common::UString &outFile,
            Aurora::GameID &game, Command &command, bool printStack, bool printControlTypes);

int main(int argc, char **argv) {
	initPlatform();

	try {
		std::vector<Common::UString> args;
		Common::Platform::getParameters(argc, argv, args);

		Aurora::GameID game = Aurora::kGameIDUnknown;

		int returnValue = 1;
		Command command = kCommandNone;
		bool printStack = false;
		bool printControlTypes = false;
		Common::UString inFile, outFile;

		if (!parseCommandLine(args, returnValue, inFile, outFile, game, command, printStack, printControlTypes))
			return returnValue;

		disNCS(inFile, outFile, game, command, printStack, printControlTypes);
	} catch (...) {
		Common::exceptionDispatcherError();
	}

	return 0;
}

bool parseCommandLine(const std::vector<Common::UString> &argv, int &returnValue,
                      Common::UString &inFile, Common::UString &outFile,
                      Aurora::GameID &game, Command &command,
                      bool &printStack, bool &printControlTypes) {
	using Common::CLI::NoOption;
	using Common::CLI::kContinueParsing;
	using Common::CLI::Parser;
	using Common::CLI::ValGetter;
	using Common::CLI::Callback;
	using Common::CLI::ValAssigner;
	using Common::CLI::makeEndArgs;
	using Common::CLI::makeAssigners;
	using Aurora::GameID;

	Common::UString encodingStr;
	NoOption inFileOpt(false, new ValGetter<Common::UString &>(inFile, "input files"));
	NoOption outFileOpt(true, new ValGetter<Common::UString &>(outFile, "output files"));
	Parser parser(argv[0], "BioWare NWScript bytecode disassembler",
	              "\nIf no output file is given, the output is written to stdout.",
	              returnValue,
	              makeEndArgs(&inFileOpt, &outFileOpt));

	parser.addSpace();
	parser.addOption("list", "Create full disassembly listing (default)", kContinueParsing,
	                 makeAssigners(new ValAssigner<Command>(kCommandListing, command)));
	parser.addOption("assembly", "Only create disassembly mnemonics", kContinueParsing,
	                 makeAssigners(new ValAssigner<Command>(kCommandAssembly, command)));
	parser.addOption("dot", "Create a graphviz dot file", kContinueParsing,
	                 makeAssigners(new ValAssigner<Command>(kCommandDot, command)));
	parser.addOption("nss", "Create a nwscript source file", kContinueParsing,
					 makeAssigners(new ValAssigner<Command>(kCommandNSS, command)));
	parser.addSpace();
	parser.addOption("nwn", "This is a Neverwinter Nights script", kContinueParsing,
	                 makeAssigners(new ValAssigner<GameID>(Aurora::kGameIDNWN, game)));
	parser.addOption("nwn2", "This is a Neverwinter Nights 2 script", kContinueParsing,
	                 makeAssigners(new ValAssigner<GameID>(Aurora::kGameIDNWN2, game)));
	parser.addOption("kotor", "This is a Knights of the Old Republic script", kContinueParsing,
	                 makeAssigners(new ValAssigner<GameID>(Aurora::kGameIDKotOR, game)));
	parser.addOption("kotor2", "This is a Knights of the Old Republic II script", kContinueParsing,
	                 makeAssigners(new ValAssigner<GameID>(Aurora::kGameIDKotOR2, game)));
	parser.addOption("jade", "This is a Jade Empire script", kContinueParsing,
	                 makeAssigners(new ValAssigner<GameID>(Aurora::kGameIDJade, game)));
	parser.addOption("witcher", "This is a The Witcher script", kContinueParsing,
	                 makeAssigners(new ValAssigner<GameID>(Aurora::kGameIDWitcher, game)));
	parser.addOption("dragonage", "This is a Dragon Age script", kContinueParsing,
	                 makeAssigners(new ValAssigner<GameID>(Aurora::kGameIDDragonAge, game)));
	parser.addOption("dragonage2", "This is a Dragon Age II script", kContinueParsing,
	                 makeAssigners(new ValAssigner<GameID>(Aurora::kGameIDDragonAge2, game)));
	parser.addSpace();
	parser.addOption("stack", "Print the stack frame for each instruction"
	                 " (Only available in list or assembly mode)",
	                 kContinueParsing,
	                 makeAssigners(new ValAssigner<bool>(true, printStack)));
	parser.addOption("control", "Print the control types for each block"
	                 " (Only available in list or assembly mode)",
	                 kContinueParsing,
	                 makeAssigners(new ValAssigner<bool>(true, printControlTypes)));
	return parser.process(argv);
}

void disNCS(const Common::UString &inFile, const Common::UString &outFile,
            Aurora::GameID &game, Command &command, bool printStack, bool printControlTypes) {

	Common::ScopedPtr<Common::SeekableReadStream> ncs(new Common::ReadFile(inFile));
	Common::ScopedPtr<Common::WriteStream> out(openFileOrStdOut(outFile));

	status("Disassembling script...");
	NWScript::Disassembler disassembler(*ncs, game);

	if (game != Aurora::kGameIDUnknown) {
		try {
			status("Analyzing script stack...");
			disassembler.analyzeStack();
		} catch (...) {
			Common::exceptionDispatcherWarnAndIgnore("Script analysis failed");
		}

		try {
			status("Analyzing control flow...");
			disassembler.analyzeControlFlow();
		} catch (...) {
			Common::exceptionDispatcherWarnAndIgnore("Control flow analysis failed");
		}
	}

	switch (command) {
		case kCommandListing:
			disassembler.createListing(*out, printStack);
			break;

		case kCommandAssembly:
			disassembler.createAssembly(*out, printStack);
			break;

		case kCommandDot:
			disassembler.createDot(*out, printControlTypes);
			break;

		case kCommandNone:
			disassembler.createListing(*out, printStack);
			break;

		case kCommandNSS:
			disassembler.createNSS(*out);
			break;

		default:
			throw Common::Exception("Invalid command %u", (uint)command);
	}

	out->flush();

	if (!outFile.empty())
		status("Disassembled \"%s\" into \"%s\"", inFile.c_str(), outFile.c_str());
}
